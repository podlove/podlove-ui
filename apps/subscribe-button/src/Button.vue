<template>
  <transition name="fade">
    <root
      v-if="state.overlay"
      v-click-outside="close"
      class="
        bg-gray-900 bg-opacity-75
        p-6
        fixed
        inset-0
        w-full
        h-full
        flex
        items-center
        justify-center
      "
    >
      <transition name="fade" mode="out-in">
        <client-screen v-if="state.listView" />
        <finish-screen v-else />
      </transition>
    </root>
  </transition>
</template>

<script>
import * as overlay from '@podlove/button-actions/overlay'
import { mapState, injectStore } from 'redux-vuex'
import Root from './components/root/Root.vue'
import ClientScreen from './screens/Clients.vue'
import FinishScreen from './screens/Finish.vue'

import * as select from './store/selectors'

export default {
  components: {
    Root,
    ClientScreen,
    FinishScreen
  },
  directives: {
    'click-outside': {
      bind(el, { value: fn }) {
        el.addEventListener('click', (evt) => {
          if (evt.target !== el) {
            return
          }

          fn()
        })
      }
    }
  },
  setup() {
    return {
      state: mapState({
        overlay: select.view.overlay,
        listView: select.view.list,
        finishView: select.view.finish,
        language: select.runtime.language
      }),
      dispatch: injectStore().dispatch
    }
  },
  computed: {
    language() {
      return this.state.language
    }
  },
  watch: {
    language() {
      this.$i18n.locale = this.language
    }
  },
  mounted() {
    this.$i18n.locale = this.language
  },
  methods: {
    close() {
      this.dispatch(overlay.hide())
    }
  }
}
</script>
