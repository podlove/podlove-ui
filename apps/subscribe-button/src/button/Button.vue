<template>
  <button id="podlove-subscribe-button" :style="filling" @click="clickevent()">
    Subscribe
  </button>
</template>

<script>
import { mapState } from 'redux-vuex'
import { selectColor, selectCover, selectFormat, selectSize, selectStyle } from 'store/selectors'

export default {
  props: {
    clickevent: {
      type: Function,
      default: null
    }
  },
  data: mapState({
    color: selectColor,
    cover: selectCover,
    format: selectFormat,
    size: selectSize,
    style: selectStyle
  }),
  computed: {
    filling() {
      if (this.style === 'outline') {
        return `border: 2px solid ${this.color}; color: ${this.color};`
      } else if (this.style === 'frameless') {
        return `color: ${this.color}; background: none;`
      } else {
        return `background: ${this.color};`
      }
    }
  }
}
</script>

<style lang="scss">
@import '~normalize-scss';
@import '~theme/reset';
@import '~theme/variable';

button {
  font-weight: 500;
  letter-spacing: 0.5px;

  text-transform: uppercase;
  transition: all 0.1s cubic-bezier(0.62, 0.28, 0.23, 0.99);
  width: 100%;
  font-size: 16px;
}
</style>
