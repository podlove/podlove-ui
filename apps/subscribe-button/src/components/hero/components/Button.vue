<template>
  <button :style="filling" @click="onClick">
    <icon type="subscribe" :size="iconSize"></icon>
    <span v-if="state.format !== 'square'">{{ $t('SUBSCRIBE') }}</span>
  </button>
</template>

<script lang="ts" setup>
import { mapState } from 'redux-vuex';
import Icon from '@podlove/components/icons';

import {
  selectColor,
  selectCover,
  selectFormat,
  selectSize,
  selectStyle
} from '../../../store/selectors';
import { computed } from 'vue';

const emits = defineEmits(['click']);

const state = mapState({
  color: selectColor,
  cover: selectCover,
  format: selectFormat,
  size: selectSize,
  style: selectStyle
});

const filling = computed(() => {
  if (state.style === 'outline') {
    return `border: 2px solid ${state.color}; color: ${state.color};`;
  } else if (state.style === 'frameless') {
    return `color: ${state.color}; background: none;`;
  } else {
    return `background: ${state.color};`;
  }
});

const iconSize = computed(() => {
  let size = 20;

  switch (state.size) {
    case 'small':
      size = 16;
      break;
    case 'medium':
      size = 18;
      break;
    case 'big':
      size = 22;
      break;
  }
  return size;
});

const onClick = () => {
  emits('click');
};
</script>

<style lang="scss"></style>
