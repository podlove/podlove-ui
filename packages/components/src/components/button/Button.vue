<script setup lang="ts">
const props = defineProps({
  href: {
    default: null,
    type: String
  },
  disabled: {
    type: Boolean,
    default: false
  }
});
</script>

<template>
  <component
    :is="href ? 'a' : 'button'"
    class="
      podlove-button
      flex
      justify-center
      h-8
      text-base
      cursor-pointer
      bg-transparent
      outline-none
      transition-opacity
      hover:opacity-75
      border
      rounded
      border-solid
      py-1
      px-2
      uppercase
      text-center
    "
    :class="{ 'opacity-50': props.disabled, 'opacity-100': !props.disabled }"
    :href="props.href"
  >
    <span class="truncate w-full h-full flex items-center justify-center">
      <slot />
    </span>
  </component>
</template>

<style lang="postcss" scoped>
.podlove-button {
  font-variant: small-caps;
  color: var(--podlove-component-button-text, var(--podlove-components-background));
  background: var(--podlove-component-button-background, var(--podlove-components-text));
  border-color: var(--podlove-component-button-border, var(--podlove-components-background));
}

.pdolove-button svg {
  display: inline;
}
</style>
