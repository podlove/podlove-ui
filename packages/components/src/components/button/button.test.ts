import { mount } from '@vue/test-utils';
import { describe, test, expect } from 'vitest';
import Button from './Button.vue';

describe('Button', () => {
  describe('component', () => {
    test('should be a Vue instance', () => {
      const wrapper = mount(Button);
      expect(wrapper.vm).toBeTruthy();
    });

    test('should provide defaults', async () => {
      const wrapper = await mount(Button, {
        slots: {
          default: '<span>Button Inner</span>'
        }
      });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('props', () => {
    const props = [
      {
        prop: 'textColor',
        value: '#fff'
      },
      {
        prop: 'backgroundColor',
        value: '#fff'
      },
      {
        prop: 'borderColor',
        value: '#fff'
      },
      {
        prop: 'href',
        value: 'http://foo.bar'
      },
      {
        prop: 'disabled',
        value: true
      }
    ];

    props.forEach(({ prop, value }) => {
      test(`should have a property for ${prop}`, async () => {
        const wrapper = await mount(Button, {
          slots: {
            default: '<span>Button Inner</span>'
          },
          props: {
            [prop]: value
          }
        });

        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });
});
