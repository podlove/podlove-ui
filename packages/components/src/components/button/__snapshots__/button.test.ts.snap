// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > component > should provide defaults 1`] = `
<button
  class="podlove-component-button flex justify-center h-8 text-base cursor-pointer bg-transparent outline-none transition-opacity hover:opacity-75 border rounded border-solid py-1 px-2 uppercase text-center opacity-100"
  data-v-585f6ee7=""
>
  <span
    class="truncate w-full h-full flex items-center justify-center"
    data-v-585f6ee7=""
  >
    
    <span>
      Button Inner
    </span>
    
  </span>
</button>
`;

exports[`Button > props > should have a property for backgroundColor 1`] = `
<button
  backgroundcolor="#fff"
  class="podlove-component-button flex justify-center h-8 text-base cursor-pointer bg-transparent outline-none transition-opacity hover:opacity-75 border rounded border-solid py-1 px-2 uppercase text-center opacity-100"
  data-v-585f6ee7=""
>
  <span
    class="truncate w-full h-full flex items-center justify-center"
    data-v-585f6ee7=""
  >
    
    <span>
      Button Inner
    </span>
    
  </span>
</button>
`;

exports[`Button > props > should have a property for borderColor 1`] = `
<button
  bordercolor="#fff"
  class="podlove-component-button flex justify-center h-8 text-base cursor-pointer bg-transparent outline-none transition-opacity hover:opacity-75 border rounded border-solid py-1 px-2 uppercase text-center opacity-100"
  data-v-585f6ee7=""
>
  <span
    class="truncate w-full h-full flex items-center justify-center"
    data-v-585f6ee7=""
  >
    
    <span>
      Button Inner
    </span>
    
  </span>
</button>
`;

exports[`Button > props > should have a property for disabled 1`] = `
<button
  class="podlove-component-button flex justify-center h-8 text-base cursor-pointer bg-transparent outline-none transition-opacity hover:opacity-75 border rounded border-solid py-1 px-2 uppercase text-center opacity-50"
  data-v-585f6ee7=""
>
  <span
    class="truncate w-full h-full flex items-center justify-center"
    data-v-585f6ee7=""
  >
    
    <span>
      Button Inner
    </span>
    
  </span>
</button>
`;

exports[`Button > props > should have a property for href 1`] = `
<a
  class="podlove-component-button flex justify-center h-8 text-base cursor-pointer bg-transparent outline-none transition-opacity hover:opacity-75 border rounded border-solid py-1 px-2 uppercase text-center opacity-100"
  data-v-585f6ee7=""
  href="http://foo.bar"
>
  <span
    class="truncate w-full h-full flex items-center justify-center"
    data-v-585f6ee7=""
  >
    
    <span>
      Button Inner
    </span>
    
  </span>
</a>
`;

exports[`Button > props > should have a property for textColor 1`] = `
<button
  class="podlove-component-button flex justify-center h-8 text-base cursor-pointer bg-transparent outline-none transition-opacity hover:opacity-75 border rounded border-solid py-1 px-2 uppercase text-center opacity-100"
  data-v-585f6ee7=""
  textcolor="#fff"
>
  <span
    class="truncate w-full h-full flex items-center justify-center"
    data-v-585f6ee7=""
  >
    
    <span>
      Button Inner
    </span>
    
  </span>
</button>
`;
